Spire agent SDS panics when using prometheus telemetry

When using the Envoy SDS and prometheus telemetry the agent panics when a client disconnects.

Similar to #937, this is due to attempting to decrement a counter at https://github.com/spiffe/spire/blob/6cde36168fcbfb9fff7ac0408cb4861471762a57/pkg/common/telemetry/agent/sdsapi.go#L16

```
	/go/pkg/mod/google.golang.org/grpc@v1.21.1/server.go:717 +0x9f
google.golang.org/grpc.(*Server).serveStreams.func1.1(0xc000c59640, 0xc0007aec00, 0x1680260, 0xc0005f6d80, 0xc001404900)
google.golang.org/grpc.(*Server).processStreamingRPC(0xc0007aec00, 0x1680260, 0xc0005f6d80, 0xc001404900, 0xc000a6a240, 0x20f24a0, 0x0, 0x0, 0x0)
	/go/pkg/mod/github.com/prometheus/client_golang@v1.0.0/prometheus/counter.go:89 +0x134
	/go/pkg/mod/google.golang.org/grpc@v1.21.1/server.go:715 +0xa1
	/go/pkg/mod/google.golang.org/grpc@v1.21.1/server.go:1282 +0xd3f
github.com/spiffe/spire/pkg/agent/endpoints/sds.(*Handler).StreamSecrets(0xc0009e00a0, 0x167d1a0, 0xc000c8c630, 0x0, 0x0)
github.com/spiffe/spire/pkg/agent/endpoints/sds.(*Handler).startCall.func1()
github.com/spiffe/spire/pkg/common/telemetry/agent.DecrSDSAPIConnectionTotalCounter(0x1680200, 0xc0009b44e0)
	/go/pkg/mod/github.com/armon/go-metrics@v0.0.0-20190430140413-ec5e00d3c878/metrics.go:81 +0x305
	/go/pkg/mod/github.com/armon/go-metrics@v0.0.0-20190430140413-ec5e00d3c878/sink.go:63 +0xbc
goroutine 263 [running]:
github.com/envoyproxy/go-control-plane/envoy/service/discovery/v2._SecretDiscoveryService_StreamSecrets_Handler(0x12be540, 0xc0009e00a0, 0x1676600, 0xc0003195c0, 0x21281a8, 0xc001404900)
github.com/spiffe/spire/pkg/common/telemetry.(*withLabels).IncrCounter(0xc0009b44e0, 0xc002c5d1c0, 0x2, 0x2, 0xc0bf800000)
github.com/armon/go-metrics.FanoutSink.IncrCounterWithLabels(0xc00015f9a0, 0x2, 0x2, 0xc003620180, 0x3, 0x4, 0x7fc4bf800000, 0xc003620240, 0x2, 0x2)
	/go/pkg/mod/github.com/armon/go-metrics@v0.0.0-20190430140413-ec5e00d3c878/prometheus/prometheus.go:192 +0x15b
created by google.golang.org/grpc.(*Server).serveStreams.func1
google.golang.org/grpc.(*Server).handleStream(0xc0007aec00, 0x1680260, 0xc0005f6d80, 0xc001404900, 0x0)
	/spire/pkg/common/telemetry/agent/sdsapi.go:16 +0x80
	/spire/pkg/common/telemetry/withlabels.go:35 +0x78
panic: counter cannot decrease in value
	/go/pkg/mod/google.golang.org/grpc@v1.21.1/server.go:1202 +0xacd
	/go/pkg/mod/github.com/envoyproxy/go-control-plane@v0.8.0/envoy/service/discovery/v2/sds.pb.go:175 +0xad
	/spire/pkg/agent/endpoints/sds/handler.go:155 +0x1260
	/spire/pkg/agent/endpoints/sds/handler.go:243 +0x4c
github.com/armon/go-metrics.(*Metrics).IncrCounterWithLabels(0xc00017a1e0, 0xc002c5d1c0, 0x3, 0x4, 0xbf800000, 0xc0008a31a0, 0x2, 0x2)
github.com/armon/go-metrics/prometheus.(*PrometheusSink).IncrCounterWithLabels(0xc000280630, 0xc003620180, 0x3, 0x4, 0xc0bf800000, 0xc003620240, 0x2, 0x2)
github.com/prometheus/client_golang/prometheus.(*counter).Add(0xc0014c2b40, 0xbff0000000000000)

	/spire/pkg/agent/endpoints/sds/handler.go:155 +0x1260
	/spire/pkg/common/telemetry/withlabels.go:35 +0x78
github.com/armon/go-metrics.FanoutSink.IncrCounterWithLabels(0xc00015f9a0, 0x2, 0x2, 0xc000a9b740, 0x3, 0x4, 0xbf800000, 0xc000a9b7c0, 0x2, 0x2)
github.com/armon/go-metrics/prometheus.(*PrometheusSink).IncrCounterWithLabels(0xc000280630, 0xc000a9b740, 0x3, 0x4, 0xc0bf800000, 0xc000a9b7c0, 0x2, 0x2)
	/go/pkg/mod/github.com/prometheus/client_golang@v1.0.0/prometheus/counter.go:89 +0x134
goroutine 314 [running]:
github.com/prometheus/client_golang/prometheus.(*counter).Add(0xc0014c2b40, 0xbff0000000000000)
panic: counter cannot decrease in value
github.com/spiffe/spire/pkg/agent/endpoints/sds.(*Handler).StreamSecrets(0xc0009e00a0, 0x167d1a0, 0xc00075cbf0, 0x0, 0x0)
	/spire/pkg/common/telemetry/agent/sdsapi.go:16 +0x80
	/go/pkg/mod/github.com/armon/go-metrics@v0.0.0-20190430140413-ec5e00d3c878/prometheus/prometheus.go:192 +0x15b
	/spire/pkg/agent/endpoints/sds/handler.go:243 +0x4c
github.com/spiffe/spire/pkg/agent/endpoints/sds.(*Handler).startCall.func1()
github.com/spiffe/spire/pkg/common/telemetry.(*withLabels).IncrCounter(0xc00093fa70, 0xc003b47e40, 0x2, 0x2, 0xc0bf800000)
github.com/armon/go-metrics.(*Metrics).IncrCounterWithLabels(0xc00017a1e0, 0xc003b47e40, 0x3, 0x4, 0xbf800000, 0xc0008ad540, 0x2, 0x2)
github.com/spiffe/spire/pkg/common/telemetry/agent.DecrSDSAPIConnectionTotalCounter(0x1680200, 0xc00093fa70)
	/go/pkg/mod/github.com/armon/go-metrics@v0.0.0-20190430140413-ec5e00d3c878/metrics.go:81 +0x305
	/go/pkg/mod/github.com/armon/go-metrics@v0.0.0-20190430140413-ec5e00d3c878/sink.go:63 +0xbc

	/go/pkg/mod/google.golang.org/grpc@v1.21.1/server.go:715 +0xa1
	/go/pkg/mod/google.golang.org/grpc@v1.21.1/server.go:717 +0x9f
google.golang.org/grpc.(*Server).handleStream(0xc0007aec00, 0x1680260, 0xc0005f6d80, 0xc001803200, 0x0)
github.com/envoyproxy/go-control-plane/envoy/service/discovery/v2._SecretDiscoveryService_StreamSecrets_Handler(0x12be540, 0xc0009e00a0, 0x1676600, 0xc000df3080, 0x21281a8, 0xc001803200)
google.golang.org/grpc.(*Server).serveStreams.func1.1(0xc000c59640, 0xc0007aec00, 0x1680260, 0xc0005f6d80, 0xc001803200)
	/go/pkg/mod/google.golang.org/grpc@v1.21.1/server.go:1282 +0xd3f
	/go/pkg/mod/google.golang.org/grpc@v1.21.1/server.go:1202 +0xacd
google.golang.org/grpc.(*Server).processStreamingRPC(0xc0007aec00, 0x1680260, 0xc0005f6d80, 0xc001803200, 0xc000a6a240, 0x20f24a0, 0x0, 0x0, 0x0)
	/go/pkg/mod/github.com/envoyproxy/go-control-plane@v0.8.0/envoy/service/discovery/v2/sds.pb.go:175 +0xad
created by google.golang.org/grpc.(*Server).serveStreams.func1
```

--- a/pkg/agent/endpoints/sds/handler.go
+++ b/pkg/agent/endpoints/sds/handler.go
@@ -8,6 +8,7 @@ import (
 	"fmt"
 	"io"
 	"strconv"
+	"sync/atomic"
 
 	api_v2 "github.com/envoyproxy/go-control-plane/envoy/api/v2"
 	auth_v2 "github.com/envoyproxy/go-control-plane/envoy/api/v2/auth"
@@ -42,6 +43,10 @@ type HandlerConfig struct {
 
 type Handler struct {
 	c HandlerConfig
+
+	// connections is a count of current connections to the API (in other words
+	// how many RPCs are outstanding) tracked for telemetry purposes.
+	connections int32
 
 	hooks struct {
 		// test hook used to synchronize receipt of a stream request
@@ -228,7 +233,7 @@ func (h *Handler) startCall(ctx context.Context) (int32, []*common.Selector, fun
 	pidStr := fmt.Sprint(pid)
 	metrics := telemetry.WithLabels(h.c.Metrics, []telemetry.Label{{Name: telemetry.SDSPID, Value: pidStr}})
 	telemetry_agent.IncrSDSAPIConnectionCounter(metrics)
-	telemetry_agent.IncrSDSAPIConnectionTotalCounter(metrics)
+	telemetry_agent.SetSDSAPIConnectionTotalGauge(metrics, atomic.AddInt32(&h.connections, 1))
 	log := h.c.Log.WithField(telemetry.SDSPID, pidStr)
 	log.Debug("Handling SDS API request")
 
@@ -238,14 +243,14 @@ func (h *Handler) startCall(ctx context.Context) (int32, []*common.Selector, fun
 	// attest some other process that happened to be assigned the original PID
 	err = watcher.IsAlive()
 	if err != nil {
-		telemetry_agent.DecrSDSAPIConnectionTotalCounter(metrics)
+		telemetry_agent.SetSDSAPIConnectionTotalGauge(metrics, atomic.AddInt32(&h.connections, -1))
 		log.Debug("Finished handling SDS API request due to error")
 		return 0, nil, nil, status.Errorf(codes.Unauthenticated, "Could not verify existence of the original caller: %v", err)
 	}
 
 	done := func() {
-		defer telemetry_agent.DecrSDSAPIConnectionTotalCounter(metrics)
-		defer log.Debug("Finished handling SDS API request")
+		telemetry_agent.SetSDSAPIConnectionTotalGauge(metrics, atomic.AddInt32(&h.connections, -1))
+		log.Debug("Finished handling SDS API request")
 	}
 
 	return pid, selectors, done, nil

--- a/pkg/common/telemetry/agent/sdsapi.go
+++ b/pkg/common/telemetry/agent/sdsapi.go
@@ -10,16 +10,9 @@ func IncrSDSAPIConnectionCounter(m telemetry.Metrics) {
 	m.IncrCounter([]string{telemetry.SDSAPI, telemetry.Connection}, 1)
 }
 
-// DecrSDSAPIConnectionTotalCounter indicate one less
-// SDS API active connection (active in that moment)
-func DecrSDSAPIConnectionTotalCounter(m telemetry.Metrics) {
-	m.IncrCounter([]string{telemetry.SDSAPI, telemetry.Connections}, -1)
-}
-
-// IncrSDSAPIConnectionTotalCounter indicate one more
-// SDS API active connections (active in that moment)
-func IncrSDSAPIConnectionTotalCounter(m telemetry.Metrics) {
-	m.IncrCounter([]string{telemetry.SDSAPI, telemetry.Connections}, 1)
+// SetSDSAPIConnectionTotalGauge sets the number of active SDS connections
+func SetSDSAPIConnectionTotalGauge(m telemetry.Metrics, connections int32) {
+	m.IncrCounter([]string{telemetry.SDSAPI, telemetry.Connections}, float32(connections))
 }
 
 // End Counters

